File: build-and-push.py
import json
import subprocess
import os
import argparse


ap = argparse.ArgumentParser(description="Build and push Docker image for BOTC Clocktower")
ap.add_argument("package_dir", default=".", nargs="?", help="Directory containing package.json (default: current directory)")
ap.add_argument("--tag-base", "-t", default="boltmk0/botc-clocktower", help="Base name for Docker image tag (default: boltmk0/botc-clocktower)")
ap.add_argument("--registry", help="Docker registry URL")
ap.add_argument("--archive-out", "-o", help="Optional path to save the built Docker image as a tar archive (e.g. botc-clocktower.tar)")

args = ap.parse_args()

package_json_path = os.path.join(args.package_dir, "package.json")
package_json_data = json.load(open(package_json_path))
version = package_json_data["version"]

latest_image_tag = f"{args.tag_base}:latest"
image_tag = f"{args.tag_base}:{version}"

if(args.registry):
    latest_image_tag = f"{args.registry}/{latest_image_tag}"
    image_tag = f"{args.registry}/{image_tag}"

print("Building and pushing Docker image:", image_tag)

# Build the Docker image
print("Building Docker image...")
subprocess.run(["docker", "build", "-t", latest_image_tag, "-t", image_tag, "--platform", "linux/amd64", args.package_dir], check=True)
# Push the Docker image to the registry
print("\n\n\n\n\nPushing Docker image...")
subprocess.run(["docker", "push", latest_image_tag], check=True)
subprocess.run(["docker", "push", image_tag], check=True)

# Optionally save the Docker image as a tar archive
if args.archive_out:
    print("\n\n\n\n\nSaving Docker image to archive:", args.archive_out)
    subprocess.run(["docker", "save", image_tag, "-o", args.archive_out], check=True)

print("\n\nDone.\n\n")

File: scripts/build-and-push.mjs
import { parseArgs } from "node:util";
import { DEFAULT_TAG_BASE, buildImage, docker, resolveTags } from "./docker-build.mjs";

const { values, positionals } = parseArgs({
    allowPositionals: true,
    options: {
        "tag-base": { type: "string", short: "t", default: DEFAULT_TAG_BASE },
        registry: { type: "string", short: "r" },
        "archive-out": { type: "string", short: "o" },
        help: { type: "boolean", short: "h" }
    }
});

if (values.help) {
    console.log(`Build and push Docker image for BOTC Clocktower

Usage: npm run docker:push -- [package_dir] [options]

  package_dir            Directory containing package.json (default: .)
  -t, --tag-base <name>  Base name for the image tag (default: ${DEFAULT_TAG_BASE})
  -r, --registry <url>   Docker registry URL
  -o, --archive-out <f>  Also save the built image as a tar archive (e.g. botc-clockmaker.tar)`);
    process.exit(0);
}

const packageDir = positionals[0] ?? ".";
const { latestImageTag, imageTag } = resolveTags({ packageDir, tagBase: values["tag-base"], registry: values.registry });

console.log("Building and pushing Docker image:", imageTag);

buildImage({ packageDir, latestImageTag, imageTag, platform: "linux/amd64" });

console.log("\n\n\n\n\nPushing Docker image...");
docker("push", latestImageTag);
docker("push", imageTag);

if (values["archive-out"]) {
    console.log("\n\n\n\n\nSaving Docker image to archive:", values["archive-out"]);
    docker("save", imageTag, "-o", values["archive-out"]);
}

console.log("\n\nDone.\n\n");
